Extend tests to cover issue #23 and conversation on PR# 24

jedesah opened the pull request:
- Adjust implementation of parsing to pass new tests
- cleanup the structure of test in `PathSpec`

jedesah commented:
The intent of this PR is to subsume #24.

I would like to harmonize the organization of `PathSpec` but I did not want to spend too much time before getting validation that it's an improvement in structure. My intention is to use the style `"""<description of test> such as "<example>""""` almost everywhere.

jedesah commented:
@mossprescott FYI, I fixed the build failure caused by a compile error with scala 2.10. Just had to remove a type annotation (refreshing change from having to add them!).

mossprescott commented on core/src/main/scala/pathy/Path.scala:
```diff
@@ -290,32 +290,26 @@ object Path {
-            FileIn(base, FileName(unescape(seg)))
-          else
-            DirIn(base, DirName(unescape(seg)))
+      val isDir = List(separator.toString, "/.", "/..").exists(str.endsWith) || str === "." || str === ".."
```
Should be `separator + ".", separator + ".."`

mossprescott commented on tests/src/test/scala/pathy/path.scala:
```diff
@@ -100,133 +84,194 @@ class PathSpecs extends Specification with ScalaCheck {
-  "parseRelFile - ./image.png" in {
-    parseRelFile("./image.png") must beSome(file("image.png"))
-  }
+      """simple file name with extension such as "image.png"""" in {
```
I'm not really sold on this concept.

For one thing, why use triple-double-quotes, when you could just pick some other way to set off the file name, say backticks?

More importantly, the phrase "such as" suggests that this is testing more than it really is. In fact it's testing parsing of exactly that one path, no more and no less. I would prefer either "simple file name with extension" (which describes the value without showing it) or even just "image.png", which would read fine I think.

jedesah replied:
I kind of like it. It conveys the intent of the test. At the same time, when looking at the spec or when running it, it's easier to parse/understand if there is an example value. I don't think it's any more misleading than if the example wasn't there. Both sound like they should test more than a single value. I am not 100% sold either, but to me it seems like the best of the three.

mossprescott commented on tests/src/test/scala/pathy/path.scala:
```diff
@@ -100,133 +84,194 @@ class PathSpecs extends Specification with ScalaCheck {
-  "parseAbsFile - /foo/image.png" in {
-    parseAbsFile("/foo/image.png") must beSome(rootDir </> dir("foo") </> file("image.png"))
-  }
+      "with a trailing /" in {
```
:+1:

jedesah replied:
Because it has no "example"? Haha, okay fine, I don't feel too strongly, so if you really prefer it this way, I can do that. This description is pretty simple though, sometimes it requires more cognitive effort to craft an example in one's head of what the description means.

mossprescott commented on tests/src/test/scala/pathy/path.scala:
```diff
@@ -100,133 +84,194 @@ class PathSpecs extends Specification with ScalaCheck {
-    parseRelDir("foo") must beNone
-  }
+      "/." in {
+        parseAbsFile("/.") must beNone
```
This is one of the fixed cases, right? It would be easier to follow if the re-org was in a separate commit, I must say.

jedesah replied:
Ya, sorry, restructuring the whole thing kind of crept up on me.

jedesah replied:
If only we had semantic diff...

mossprescott commented on tests/src/test/scala/pathy/path.scala:
```diff
@@ -100,133 +84,194 @@ class PathSpecs extends Specification with ScalaCheck {
-  "parseRelFile - image.png" in {
-    parseRelFile("image.png") must beSome(file("image.png"))
-  }
+  "parseRelFile" >> {
```
As a general comment, I think it would be good to make it clearer that all of these tests apply to `posixCodex` specifically. How about this?
```scala
"posixCodec" >> {
  import posixCodec._

  "parseRelFile" >> {
    "should X" in {
```

jedesah replied:
Agreed. Although I think some of them are meant to be general. For those I could make an `Arbitrary[PathCodec]` and make them properties. Unfortunately, `scalacheck` is not smart enough to realize that there are only a finite amount of `PathCodec` and only test them once... *sigh*

mossprescott replied:
If that's the case, I think you want some kind of `testAll` rather than a generator.

But I doubt that there are really too many tests that are really common. As soon as you have a `/` you work only for posixCodec, right?

jedesah replied:
Ya, I take back my comment on a `Arbitrary[PathCoded]`. I think the tests are mostly meant to test `placeholder` and `/` is easier than `\\`.

mossprescott commented on tests/src/test/scala/pathy/PrintPathSpecs.scala:
```diff
@@ -0,0 +1,44 @@
+import org.specs2.mutable.Specification
+import pathy.Path._
+
+class PrintPathSpecs extends Specification {
```
It's not really clear to me why these tests are broken out here while the "placeholder" tests, which test the same method, are still in the other file.

jedesah replied:
Ah, I thought these were tests for printing, but that is not clear to me anymore. Happy to put them back where they were. But that other file is getting pretty big, we should think about splitting it up a bit.

jedesah commented on tests/src/test/scala/pathy/path.scala:
```diff
@@ -27,28 +27,22 @@ class PathSpecs extends Specification with ScalaCheck {
-
-  "print and parse again should produce same Path" ! prop { path: AbsFile[Sandboxed] =>
-    parseAbsFile(printPath(path)).get must_== path
+  "print and parse again should produce same Path" >> {
```
@mossprescott It bugged me that the hard coding of `/` was only detected during code review as opposed to detected by some tests. I was able to expose the bug through tests with this. :smile:

jedesah commented:
@mossprescott I think I addressed your concerns about how the tests are structured. I tried not to go too overboard and just make an incremental improvement. The diff is not the best, but hopefully the improved structure will make it easier to follow.

jedesah commented:
Ya, because I renames `path.scala` to `PathSpecs.scala`, it's a little harder to follow, but, as briefly mentioned in the commit message, the only modification I made to tests in `PathSpecs.scala` vs `path.scala` is removing the use of `unsafePrintPath` from the `canonicalize` `renameFile` and `sandbox` tests.

mossprescott commented on scalacheck/src/main/scala/pathy/scalacheck/PathyArbitrary.scala:
```diff
@@ -40,6 +40,9 @@ object PathyArbitrary {
     Arbitrary(genDirName[RandomSeg])
 
+  implicit val arbitraryPathCodec: Arbitrary[PathCodec] =
+    Arbitrary(Gen.oneOf(posixCodec,windowsCodec))
```
Seems like this is unused now. I don't think we need it anymore, now that you have validateIsLossless.

jedesah replied:
Okay, will remove it.

mossprescott commented:
Just the one comment on the unused Arbitrary.

Looks good. Looking forward to getting this merged and using it to test our PathUriCodec, too.
